spec/acceptance/datasets_spec.rb

```diff
@@ -0,0 +1,215 @@
+require 'rails_helper'
+require 'rspec_api_documentation/dsl'
+require 'helpers/acceptance_spec_helper'
+
+def id_params
+  parameter :id, 'Dataset id in request url', required: true
+end
+
+def body_params
+  parameter :name, 'Name of dataset', scope: :dataset, :required => true
+  parameter :description, 'Description of dataset', scope: :dataset
+end
+
+# https://github.com/zipmark/rspec_api_documentation
+resource 'Datasets' do
+
+  header 'Accept', 'application/json'
+  header 'Content-Type', 'application/json'
+  header 'Authorization', :authentication_token
+
+  let(:format) { 'json' }
+
+  create_entire_hierarchy
+
+  # Create post parameters from factory
+  let(:post_attributes) { FactoryGirl.attributes_for(:dataset, name: 'New Dataset name') }
+
+  ################################
+  # INDEX
+  ################################
+
+  # Expected count is 2 because of the 'default dataset', which is seed data added during a clean install.
+
+  get '/datasets' do
+    let(:authentication_token) { admin_token }
+    standard_request_options(:get, 'INDEX (as admin)', :ok, {expected_json_path: 'data/0/name', data_item_count: 2})
+  end
+
+  # will only return dataset that are created by the user
+
+  get '/datasets' do
+    let(:authentication_token) { reader_token }
+    standard_request_options(:get, 'INDEX (as non admin user)', :ok, {response_body_content: '200', data_item_count: 0})
+  end
+
+  get '/datasets' do
+    let(:authentication_token) { invalid_token }
+    standard_request_options(:get, 'INDEX (with invalid token)', :unauthorized, {expected_json_path: get_json_error_path(:sign_up)})
+  end
+
+  get '/datasets' do
+    standard_request_options(:get, 'INDEX (as anonymous user)', :ok, {remove_auth: true, response_body_content: '200', data_item_count: 0})
+  end
+
+  ################################
+  # CREATE
+  ################################
+
+  post '/datasets' do
+    body_params
+    let(:raw_post) { {'dataset' => post_attributes}.to_json }
+    let(:authentication_token) { admin_token }
+    standard_request_options(:post, 'CREATE (as admin)', :created, {expected_json_path: 'data/name', response_body_content: 'New Dataset name'})
+  end
+
+  post '/datasets' do
+    body_params
+    let(:raw_post) { {'dataset' => post_attributes}.to_json }
+    let(:authentication_token) { reader_token }
+    standard_request_options(:post, 'CREATE (as non admin user)', :created, {expected_json_path: 'data/name', response_body_content: 'New Dataset name'})
```

Comment:
please test all well-known users (`harvester`, `owner`, and `writer`) - we've had permissions bugs before where some users weren't tested out of convenience and of course bugs were found

This comment applies to other tests in this file

Reply:
Fixed